internal/state/storage/storage.go

```diff
@@ -0,0 +1,392 @@
+	}
+
+	duration := time.Since(start)
+
+	storage.logger.Info("state saved", zap.Int("clusters", stats.NumClusters), zap.Int("affiliates", stats.NumAffiliates),
+		zap.Int("endpoints", stats.NumEndpoints), zap.Duration("duration", duration))
+
+	storage.operationsMetric.WithLabelValues(operationSave, statusSuccess).Inc()
+	storage.lastSnapshotSizeMetric.WithLabelValues(operationSave).Set(float64(stats.Size))
+	storage.lastOperationClustersMetric.WithLabelValues(operationSave).Set(float64(stats.NumClusters))
+	storage.lastOperationAffiliatesMetric.WithLabelValues(operationSave).Set(float64(stats.NumAffiliates))
+	storage.lastOperationEndpointsMetric.WithLabelValues(operationSave).Set(float64(stats.NumEndpoints))
+	storage.lastOperationDurationMetric.WithLabelValues(operationSave).Set(duration.Seconds())
+
+	return nil
+}
+
+// Load loads all clusters' states from the persistent storage.
+func (storage *Storage) Load() (err error) {
+	defer func() {
+		if err != nil {
+			storage.operationsMetric.WithLabelValues(operationLoad, statusError).Inc()
+		}
+	}()
+
+	start := time.Now()
+
+	// open file for reading
+	file, err := os.Open(storage.path)
+	if err != nil {
+		return fmt.Errorf("failed to open file: %w", err)
+	}
+
+	defer file.Close() //nolint:errcheck
+
+	stats, err := storage.Import(file)
+	if err != nil {
+		return fmt.Errorf("failed to read snapshot: %w", err)
+	}
+
+	if err = file.Close(); err != nil {
+		return fmt.Errorf("failed to close file: %w", err)
+	}
+
+	duration := time.Since(start)
+
+	storage.logger.Info("state loaded", zap.Int("clusters", stats.NumClusters), zap.Int("affiliates", stats.NumAffiliates),
+		zap.Int("endpoints", stats.NumEndpoints), zap.Duration("duration", duration))
+
+	storage.operationsMetric.WithLabelValues(operationLoad, statusSuccess).Inc()
+	storage.lastSnapshotSizeMetric.WithLabelValues(operationLoad).Set(float64(stats.Size))
+	storage.lastOperationClustersMetric.WithLabelValues(operationLoad).Set(float64(stats.NumClusters))
+	storage.lastOperationAffiliatesMetric.WithLabelValues(operationLoad).Set(float64(stats.NumAffiliates))
+	storage.lastOperationEndpointsMetric.WithLabelValues(operationLoad).Set(float64(stats.NumEndpoints))
+	storage.lastOperationDurationMetric.WithLabelValues(operationLoad).Set(duration.Seconds())
+
+	return nil
+}
+
+// Import imports all clusters' states from the given reader.
+func (storage *Storage) Import(reader io.Reader) (SnapshotStats, error) {
+	const bufferSize = 32
+
+	size := 0
+	numClusters := 0
+	numAffiliates := 0
+	numEndpoints := 0
+
+	index := 0
+	buffer := make([]byte, bufferSize)
+
+	// unmarshal the clusters in a streaming manner and import them into the state
+	if err := storage.state.ImportClusterSnapshots(func() (*storagepb.ClusterSnapshot, bool, error) {
+		for {
+			numRead, err := reader.Read(buffer[index:])
+			size += numRead
+
+			if err != nil {
+				if err == io.EOF { //nolint:errorlint
```

Comment:
there is a slight problem here - as we  can come here both when we're at the middle of the cluster (when reading more) and when we are in between reads, one of the EOFs is fine and another is unexpected.

I would instead rework this code to simplify it:


* use `bufio.Reader` to avoid hitting actual reader with many reads
* read enough bytes to read the protobuf type/tag/size, decode actual size and more bytes to read 
* read exactly required amount of bytes, and pass it to decoder

Basically split reading header which has fixed max size and other (unknown) size part into two separate reads, but amortized via `bufio.Reader`.

Use `io.ReadFull` to make sure we can have full buffer, no matter what Reader returns.